File: langs/swift/Tests/VMxTests/OwnedResourceConformanceTests.swift
import Combine
import XCTest
@testable import VMx

private enum CleanupFailure: Error { case boom }

private final class OwnedProbeVM: ComponentVMBase {
    private let disposeHook: (() -> Void)?
    override var type: ViewModelType { .component }

    init(hub: MessageHubProtocol, disposeHook: (() -> Void)? = nil) {
        self.disposeHook = disposeHook
        super.init(
            name: "probe",
            hub: hub,
            dispatcher: ImmediateDispatcher.INSTANCE
        )
    }

    func register(_ cleanup: @escaping () throws -> Void) {
        own(cleanup)
    }

    func register(_ cancellable: any Cancellable) {
        own(cancellable)
    }

    override func _onDispose() {
        disposeHook?()
    }
}

final class OwnedResourceConformanceTests: XCTestCase {
    /// DISP-007 — owned resources are cleaned in LIFO order.
    func testDisp007OwnedResourcesAreLifo() {
        var trace: [String] = []
        let vm = OwnedProbeVM(hub: MessageHub()) { trace.append("hook") }
        vm.register { trace.append("closure") }
        vm.register(AnyCancellable { trace.append("cancellable") })
        vm.register { trace.append("last") }

        vm.dispose()

        XCTAssertEqual(trace, ["hook", "last", "cancellable", "closure"])
    }

    /// DISP-008 — repeated VM disposal cleans each resource exactly once.
    func testDisp008RepeatedDisposeCleansOnce() {
        var calls = 0
        let vm = OwnedProbeVM(hub: MessageHub())
        vm.register { calls += 1 }
        vm.dispose()
        vm.dispose()
        XCTAssertEqual(calls, 1)
    }

    /// DISP-009 — cleanup failure is swallowed and later resources still run.
    func testDisp009CleanupFailureIsIsolated() {
        var trace: [String] = []
        let vm = OwnedProbeVM(hub: MessageHub())
        vm.register { trace.append("first") }
        vm.register { throw CleanupFailure.boom }
        vm.register { trace.append("last") }

        vm.dispose()

        XCTAssertEqual(trace, ["last", "first"])
    }

    /// DISP-010 — registration after disposal is cleaned immediately once.
    func testDisp010PostDisposeRegistrationCleansImmediately() {
        var calls = 0
        let vm = OwnedProbeVM(hub: MessageHub())
        vm.dispose()
        vm.register { calls += 1 }
        vm.dispose()
        XCTAssertEqual(calls, 1)
    }

    /// DISP-011 — disposal-lifetime resources survive reconstruct.
    func testDisp011OwnedResourceSurvivesReconstruct() throws {
        var calls = 0
        let vm = OwnedProbeVM(hub: MessageHub())
        vm.register { calls += 1 }
        try vm.construct()
        try vm.reconstruct()
        XCTAssertEqual(calls, 0)
        vm.dispose()
        XCTAssertEqual(calls, 1)
    }

    /// DISP-012 — injected hub is publicly visible and read-only.
    func testDisp012HubIsPubliclyVisible() {
        let hub = MessageHub()
        let vm = OwnedProbeVM(hub: hub)
        XCTAssertTrue(vm.hub === hub)
    }

    /// DISP-013 — VM disposal does not dispose the shared injected hub.
    func testDisp013VmDoesNotOwnHub() {
        let hub = MessageHub()
        let vm = OwnedProbeVM(hub: hub)
        var received = 0
        let cancellable = hub.messages.sink { _ in received += 1 }
        vm.dispose()
        let baseline = received

        hub.send(ConstructionStatusChangedMessage(
            sender: vm,
            senderName: vm.name,
            status: vm.status
        ))

        XCTAssertEqual(received, baseline + 1)
        cancellable.cancel()
    }
}

File: langs/csharp/tests/VMx.Conformance.Tests/OwnedResourceConformanceTests.cs
using System.Reactive.Disposables;
using System.Reactive.Linq;
using FluentAssertions;
using VMx.Components;
using VMx.Messages;
using VMx.Services;
using VMx.Tests.Helpers;
using Xunit;

namespace VMx.Conformance.Tests;

/// <summary>Owned-resource and public-hub conformance (DISP-007..013).</summary>
public class OwnedResourceConformanceTests
{
    [Fact, Trait("Conformance", "DISP-007")]
    public void DISP_007_Owned_Resources_Are_Cleaned_In_Lifo_Order()
    {
        var trace = new List<string>();
        var vm = new ProbeVM(new TestHub(), () => trace.Add("hook"));
        vm.Register(() => trace.Add("action"));
        vm.Register(Disposable.Create(() => trace.Add("disposable")));
        vm.Register(() => trace.Add("last"));

        vm.Dispose();

        trace.Should().Equal("hook", "last", "disposable", "action");
    }

    [Fact, Trait("Conformance", "DISP-008")]
    public void DISP_008_Repeated_Dispose_Cleans_Each_Resource_Once()
    {
        var count = 0;
        var vm = new ProbeVM(new TestHub());
        vm.Register(() => count++);

        vm.Dispose();
        vm.Dispose();

        count.Should().Be(1);
    }

    [Fact, Trait("Conformance", "DISP-009")]
    public void DISP_009_Cleanup_Failure_Is_Swallowed_And_Does_Not_Stop_Later_Resources()
    {
        var trace = new List<string>();
        var vm = new ProbeVM(new TestHub());
        vm.Register(() => trace.Add("first"));
        vm.Register(() => throw new InvalidOperationException("boom"));
        vm.Register(() => trace.Add("last"));

        var act = vm.Dispose;

        act.Should().NotThrow();
        trace.Should().Equal("last", "first");
    }

    [Fact, Trait("Conformance", "DISP-010")]
    public void DISP_010_Registration_After_Dispose_Cleans_Immediately_Once()
    {
        var count = 0;
        var vm = new ProbeVM(new TestHub());
        vm.Dispose();

        vm.Register(() => count++);
        vm.Dispose();

        count.Should().Be(1);
    }

    [Fact, Trait("Conformance", "DISP-011")]
    public void DISP_011_Owned_Resources_Survive_Reconstruct_Until_Final_Dispose()
    {
        var count = 0;
        var vm = new ProbeVM(new TestHub());
        vm.Register(() => count++);
        vm.Construct();

        vm.Reconstruct();

        count.Should().Be(0);
        vm.Dispose();
        count.Should().Be(1);
    }

    [Fact, Trait("Conformance", "DISP-012")]
    public void DISP_012_Injected_Hub_Is_Publicly_Read_Only()
    {
        var hub = new TestHub();
        var vm = new ProbeVM(hub);

        vm.Hub.Should().BeSameAs(hub);
    }

    [Fact, Trait("Conformance", "DISP-013")]
    public void DISP_013_VM_Disposal_Does_Not_Dispose_Injected_Hub()
    {
        var hub = new MessageHub();
        var vm = new ProbeVM(hub);
        var received = 0;
        using var subscription = hub.Messages.Subscribe(_ => received++);
        vm.Dispose();
        var baseline = received;

        hub.Send(ConstructionStatusChangedMessage.Create(vm, vm.Name, vm.Status));

        received.Should().Be(baseline + 1);
    }

    private sealed class ProbeVM : ComponentVMBase
    {
        private readonly Action? _onDispose;

        public ProbeVM(IMessageHub hub, Action? onDispose = null)
            : base("probe", "", hub, new TestDispatcher(), null, null)
        {
            _onDispose = onDispose;
        }

        public override ViewModelType Type => ViewModelType.Component;

        public void Register(Action cleanup) => Own(cleanup);

        public void Register<T>(T resource) where T : IDisposable => Own(resource);

        protected override void OnDispose() => _onDispose?.Invoke();
    }
}
